aesd-char-driver/main.c

```diff
@@ -172,22 +178,105 @@ static ssize_t aesd_write(struct file *filp, const char __user *buf, size_t coun
         }
     }
     
+    //*f_pos += bytesToAdd;
     retval = bytesToAdd;
+    
 
     endfunc:
     //release semaphore
     up(device->aesd_dev_lock);
-    PDEBUG("wrote %zu bytes",bytesToAdd);
-    aesd_print();
+    PDEBUG("Exiting write:\tf_pos:%lld\t\tret:%lu",*f_pos,retval);
+    //aesd_print();
     return retval;
 }
 
+loff_t aesd_llseek(struct file *filp, loff_t offset, int whence)
+{
+    //PDEBUG("\n\n\n\n\n\n\n");
+    PDEBUG("Calling llseek:\toffset: %lld\tf_pos:%lld",offset,filp->f_pos);
+    
+    struct aesd_dev* device = (struct aesd_dev*) filp->private_data;
+    loff_t returnedOffset = 0;
+    //acquire semaphore
+    down(device->aesd_dev_lock);
+
+    //Check circular buffer length
+    loff_t bufferLength = 0;
+    
+    //this only works because we never actually remove anything from the buffer
+    for (int i = 0; i < AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED; i++){
+        bufferLength += device->circularBuffer->entry[i].size;
+    }
+
+    //release semaphore
+    up(device->aesd_dev_lock);
+
+    returnedOffset = fixed_size_llseek(filp, offset, whence, bufferLength);
+
+    PDEBUG("Exiting llseek:\tf_pos:%lld\t\tret:%lld",filp->f_pos,returnedOffset);
+    return returnedOffset;
+}
+
+long aesd_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
+{
+    //PDEBUG("\n\n\n\n\n\n\n");
+    PDEBUG("Calling ioctl:\tcmd: %u\tf_pos: %lld", cmd, filp->f_pos);
+    //PDEBUG("Expecting:\tcmd: %lu\tgot: %u", AESDCHAR_IOCSEEKTO, cmd);
+    switch (cmd) {
+    case AESDCHAR_IOCSEEKTO: 
+        PDEBUG("Received ioctl_AESDCHAR_IOCSEEKTO command");
+        struct aesd_dev *device = (struct aesd_dev *)filp->private_data;
+        struct aesd_seekto seekto;
+        loff_t byteCount = 0;
+
+        if (copy_from_user(&seekto, (const void __user *)arg, sizeof(seekto))){
+            PDEBUG("Failed to copy arguments");
+            return -EFAULT;
+        }
+            
+        PDEBUG("Arguments:\twrite_cmd: %u\twrite_cmd_offset:%u", seekto.write_cmd, seekto.write_cmd_offset);
+
+        if (seekto.write_cmd >= AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED) {
+            // Circular Buffer not large enough to store that many commands
+            return -EINVAL;
+        }
+
+        ssize_t startIndex = device->circularBuffer->out_offs;
+        ssize_t endIndex = (device->circularBuffer->out_offs + seekto.write_cmd) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+
+        // Acquire semaphore
+        down(device->aesd_dev_lock);
+
+        for (int i = startIndex; i != (endIndex + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED; i = (i + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
+        {
+            if (i == endIndex) {
+                byteCount += seekto.write_cmd_offset;
+            } else {
+                byteCount += device->circularBuffer->entry[i].size;
+            }
+        }
+
+        up(device->aesd_dev_lock);
+
+        long retval = aesd_llseek(filp, byteCount, SEEK_SET);
```

Comment:
I like the idea to reuse logic with llseek but I think in this case it might be better to make the ioctl and llseek use a shared helper function rather than having ioctl attmempt to call llseek directly.